python/ray/util/spark/databricks_hook.py

```diff
@@ -63,12 +73,69 @@ def on_ray_dashboard_created(self, port):
         )
 
-    def on_spark_background_job_created(self, job_group_id):
+    def on_cluster_created(self, ray_cluster_handler):
+        db_api_entry = get_dbutils().entry_point
         try:
-            get_dbutils().entry_point.registerBackgroundSparkJobGroup(job_group_id)
+            db_api_entry.registerBackgroundSparkJobGroup(
+                ray_cluster_handler.spark_job_group_id
+            )
         except Exception:
             _logger.warning(
                 "Register ray cluster spark job as background job failed. You need to "
                 "manually call `ray_cluster_on_spark.shutdown()` before detaching "
                 "your databricks python REPL."
             )
+
+        auto_shutdown_timeout_minutes = float(
+            os.environ.get(DATABRICKS_RAY_ON_SPARK_AUTOSHUTDOWN_TIMEOUT_MINUTES, "30")
+        )
+        if auto_shutdown_timeout_minutes == 0:
+            _logger.info(
+                "The Ray cluster will keep running until you manually detach the "
+                "databricks notebook or call "
+                "`ray.util.spark.shutdown_ray_cluster()`."
+            )
+        else:
+            _logger.info(
+                "The Ray cluster will be shut down automatically if you don't run "
+                "commands on the databricks notebook for "
+                f"{auto_shutdown_timeout_minutes} minutes. You can change the "
+                "timeout minutes by setting "
+                "'DATABRICKS_RAY_ON_SPARK_AUTOSHUTDOWN_TIMEOUT_MINUTES' environment "
+                "variable, setting it to 0 means infinite timeout."
+            )
+
+        def auto_shutdown_watcher():
+            auto_shutdown_timeout_millis = auto_shutdown_timeout_minutes * 60 * 1000
+            while True:
+                if ray_cluster_handler.is_shutdown:
+                    # The cluster is shut down. The watcher thread exits.
+                    return
+
+                try:
+                    idle_time = (
+                        db_api_entry.getIdleTimeMillisSinceLastNotebookExecution()
+                    )
+                except Exception:
+                    _logger.warning(
+                        "Your current Databricks Runtime version does not support API "
+                        "`getIdleTimeMillisSinceLastNotebookExecution`, we cannot "
+                        "automatically shut down Ray cluster when databricks notebook "
+                        "is inactive, you need to manually detach databricks notebook "
+                        "or call `ray.util.spark.shutdown_ray_cluster()` to shut down "
+                        "Ray cluster on spark."
+                    )
+                    return
+
+                if idle_time > auto_shutdown_timeout_millis:
+                    from ray.util.spark import cluster_init
+
+                    with cluster_init._active_ray_cluster_rwlock:
+                        if ray_cluster_handler is cluster_init._active_ray_cluster:
```

Comment:
I add lock for `setup_ray_cluster` and `shutdown_ray_cluster`,
but when code runs to here, it is possible that user already manually shut down current active cluster and created another cluster, so I add `ray_cluster_handler is cluster_init._active_ray_cluster` condition check here.